pageserver/src/pgdatadir_mapping.rs

```diff
@@ -490,12 +491,35 @@ impl Timeline {
         if !dbdirs.contains_key(&(tag.spcnode, tag.dbnode)) {
             return Ok(false);
         }
-        // fetch directory listing
+
+        // Read path: first read the new reldir keyspace, and then
+        // merge the result with the old reldir key.
+        // TODO: if IndexPart::rel_size_migration is `Migrated`, we only need to read from v2.
+
+        if self.get_rel_size_v2_enabled() {
+            // fetch directory listing (new)
+            let key = rel_tag_sparse_key(tag.spcnode, tag.dbnode, tag.relnode, tag.forknum);
+            let buf = RelDirExists::decode_option(version.sparse_get(self, key, ctx).await?)
+                .ok_or_else(|| {
+                    PageReconstructError::Other(anyhow::anyhow!("invalid reldir key"))
+                })?;
+            let exists_v2 = buf == RelDirExists::Exists;
+            // Fast path: if the relation exists in the new format, return true.
+            // TODO: we should have a verification mode that checks both keyspaces
+            // to ensure the relation only exists in one of them.
+            if exists_v2 {
+                return Ok(true);
+            }
```

Comment:
Can't we return early here in the `RelDirExists::Removed` case too? If a key is present, it is always authoritative, right?

Reply:
Yes, we can return early if the value is empty. But `RelDirExists::Removed` corresponds to two cases: the value is empty or the value does not exist. An empty sparse value will be removed during image layer creation. So I'd rather keep it simple here.

Reply:
Oh, I see. Yeah, makes sense.